cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java

```diff
@@ -187,6 +210,54 @@ public void run(RuntimeJobEnvironment runtimeJobEnv) throws Exception {
     Deque<Service> coreServices = createCoreServices(injector, systemArgs, cluster);
     startCoreServices(coreServices);
 
+    // regenerate app spec
+    ConfiguratorFactory configuratorFactory = injector.getInstance(ConfiguratorFactory.class);
+
+    try {
+      Map<String, String> systemArguments = new HashMap<>(programOpts.getArguments().asMap());
+      File pluginDir = new File(programOpts.getArguments().getOption(ProgramOptionConstants.PLUGIN_DIR,
+                                                                     DistributedProgramRunner.PLUGIN_DIR));
+      // create a directory to store plugin artifacts if the regeneration of app spec fetches some plugin artifact
+      if (!pluginDir.exists()) {
```

Comment:
Should this directory be deleted after execution?

Reply:
No, this if condition actually cannot happen for our pipeline since our pipeline always contain some plugins, thus this plugin dir will exist before we get there. We do not delete this folder, this will help us to debug if something went wrong. The entire run folder will get deleted if run successful so we don't need to worry about it.